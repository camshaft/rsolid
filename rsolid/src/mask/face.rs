use super::*;

pub fn cylinder(radius: impl Into<Length>, shape: impl IntoObject<2>) -> Object {
    let shape = shape.into_object() >> right(radius);
    shape >> rotate_extrude()
}

#[test]
fn cylinder_test() {
    assert_3d_snapshot!(cylinder(100.0, fillet(10.0)));
}

pub fn cube(width: f64, height: f64, shape: impl IntoObject<2>) -> Object {
    let shape = &shape.into_object();
    let a = edge(height, shape) >> xrot(90);
    let b = edge(width, shape) >> rotate([90, 0, 90]);
    let a1 = &a >> right(width / 2.0);
    let a2 = &a1 >> mirror([1, 0, 0]);
    let b1 = &b >> fwd(height / 2.0);
    let b2 = &b1 >> mirror([0, 1, 0]);

    a1 + a2 + b1 + b2
}

#[test]
fn cube_test() {
    assert_3d_snapshot!(cube(50.0, 100.0, fillet(10.0)));
}
