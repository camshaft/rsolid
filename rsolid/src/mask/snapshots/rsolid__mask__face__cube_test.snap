---
source: rsolid/src/mask/face.rs
expression: "cube(50.0, 100.0, fillet(10.0))"
---
function _v0() = true;
function _v1() = 10;
function _v2() = [_v1(), _v1()];
module _v3() { square(center=_v0(), size=_v2()); }
module _v4() { circle(r=_v1()); }
module _v5 (y=undef) { translate([0, -y, 0]) children(); }
function _v6() = 5;
module _v7() { _v5(y=_v6()) children(); }
module _v8() { _v7() _v4(); }
module _v9 (x=undef) { translate([-x, 0, 0]) children(); }
module _v10() { _v9(x=_v6()) children(); }
module _v11() { _v10() _v8(); }
module _v12() { difference() { _v3(); _v11(); }; }
function _v13() = 0.01;
function _v14() = [_v1(), _v13()];
module _v15() { square(center=_v0(), size=_v14()); }
module _v16 (y=undef) { translate([0, y, 0]) children(); }
module _v17() { _v16(y=_v6()) children(); }
module _v18() { _v17() _v15(); }
function _v19() = [_v13(), _v1()];
module _v20() { square(center=_v0(), size=_v19()); }
module _v21 (x=undef) { translate([x, 0, 0]) children(); }
module _v22() { _v21(x=_v6()) children(); }
module _v23() { _v22() _v20(); }
module _v24() { union() { _v18(); _v23(); }; }
module _v25 () { if($preview) { children(); } }
module _v26() { _v25() _v24(); }
module _v27() { union() { _v12(); _v26(); }; }
module _v28() { _v7() _v27(); }
module _v29() { _v10() _v28(); }
function _v30() = 100;
module _v31() { linear_extrude(center=_v0(), height=_v30()) children(); }
module _v32() { _v31() _v29(); }
module _v33 (a=0) { rotate([a, 0, 0]) children(); }
function _v34() = 90;
module _v35() { _v33(a=_v34()) children(); }
module _v36() { _v35() _v32(); }
function _v37() = 25;
module _v38() { _v21(x=_v37()) children(); }
module _v39() { _v38() _v36(); }
function _v40() = 1;
function _v41() = 0;
function _v42() = [_v40(), _v41(), _v41()];
module _v43() { mirror(v=_v42()) children(); }
module _v44() { _v43() _v39(); }
module _v45() { union() { _v39(); _v44(); }; }
function _v46() = 50;
module _v47() { linear_extrude(center=_v0(), height=_v46()) children(); }
module _v48() { _v47() _v29(); }
function _v49() = [_v34(), _v41(), _v34()];
module _v50() { rotate(a=_v49()) children(); }
module _v51() { _v50() _v48(); }
module _v52() { _v16(y=_v46()) children(); }
module _v53() { _v52() _v51(); }
module _v54() { union() { _v45(); _v53(); }; }
function _v55() = [_v41(), _v40(), _v41()];
module _v56() { mirror(v=_v55()) children(); }
module _v57() { _v56() _v53(); }
module _v58() { union() { _v54(); _v57(); }; }

_v58();
